worker/jobs/bulkUpsertSamplesJob.js

```diff
@@ -41,26 +42,32 @@ module.exports = (job, done) => {
   }
 
   const dbStartTime = Date.now();
+  let dbEndTime;
   cacheSampleModel.bulkUpsertByName(samples, user, readOnlyFields)
     .then((results) => {
-      const dbEndTime = Date.now();
-      let errorCount = 0;
+      upsertResults = results;
+      dbEndTime = Date.now();
+      const publishPromises = [];
 
       /*
        * count failed promises and send the good samples to the client by
        * publishing it to the redis channel
        */
-      for (let i = 0; i < results.length; i++) {
-        if (results[i].isFailed) {
+      for (let i = 0; i < upsertResults.length; i++) {
+        if (upsertResults[i].isFailed) {
           errorCount++;
 
           // we just need "explanation" to be added to the errors
-          errors.push(results[i].explanation);
+          errors.push(upsertResults[i].explanation);
         } else {
-          publisher.publishSample(results[i], subHelper.model);
+          publishPromises
+            .push(publisher.publishSample(upsertResults[i], subHelper.model));
         }
       }
 
+      return Promise.all(publishPromises);
```

Comment:
Makes perfect sense. Glad we finally figured this out!

I think we can simplify this with something like:
      Promise.all(results.map((result) => {
        if (result.isFailed) {
          errors.push(...);
          return Promise.resolve();
        } else {
          return publisher.publishSample(...)
        }
      });

Reply:
All credit goes to @shriramshankar on the "figuring out" side :)